### PR TITLE
enhancement(regex_parser transform): Add RegexSet support to regex

### LINKED ISSUE #2469: Add RegexSet support to regex
This is a more specific form of #1477. (and thus related to #2418)

Add `RegexSet` support to the `regex_parser` transform.

The docs for RegexSet are: https://docs.rs/regex/1.3.7/regex/struct.RegexSet.html

This would allow users to match over multiple regexes efficiently and save them some configuration woes.

```toml
[transforms.<transform-id>]
  type = "regex_parser"
  field = "message"
  regexes = [
     '^(?P<host>[\w\.]+) - (?P<user>[\w]+) (?P<bytes_in>[\d]+) \[(?P<timestamp>.*)\] "(?P<method>[\w]+) (?P<path>.*)" (?P<status>[\d]+) (?P<bytes_out>[\d]+)$',
     '^(?P<host>[\w\.]+) - (?P<user>[\w]+) (?P<bytes_in>[\d]+) \[(?P<timestamp>.*)\] "(?P<method>[\w]+) (?P<path>.*)" (?P<status>[\d]+) (?P<bytes_out>[\d]+)$',
     '^(?P<host>[\w\.]+) - (?P<user>[\w]+) (?P<bytes_in>[\d]+) \[(?P<timestamp>.*)\] "(?P<method>[\w]+) (?P<path>.*)" (?P<status>[\d]+) (?P<bytes_out>[\d]+)$',
  ]

[transforms.<transform-id>.types]
  bytes_in = "int"
  timestamp = "timestamp|%d/%m/%Y:%H:%M:%S %z"
  status = "int"
  bytes_out = "int"
```

### PR DESCRIPTION
**Pull request by mre**
This allows to specify multiple regular expressions to be defined
that will be matched on the input using `regex::RegexSet`.
Fixes #2469.

Note that this is a **breaking change**, as it requires all configs to be rewritten
to use `regexes = ["..."]` instead of `regex = "..."`. Alternatively, we could
support both fields in the config; although I think this would lead to confusion down
the road.

## Design decisions

- The individual regular expressions are kept in a vector named `regexes`, while
  the actual `RegexSet` is called `regexset` in the transform. The reason for keeping both is that
  [RegexSet::matches] returns the set of regular expressions that match in the given text.
  The set returned contains the index of each regular expression that matches in the given text.
  The index is in correspondence with the order of regular expressions given to 
  RegexSet's constructor, which matches with the index in the vector.
- Type coercions are applied to all captures with the same name across all patterns.
  The reasoning behind this was that types with the same name will probably be
  used similarly further down the pipeline, no matter which pattern was matching.
  Also, splitting up the coercions for each pattern would likely cause additional duplication
  and make the configuration syntax harder to understand for beginners.

## Open questions

- If possible, we should avoid the two calls to `clone()`. Was considering to store references
  to the regular expressions inside the vector instead of the expressions themselves,
  but I was hoping to avoid lifetimes to keep the code more maintainable. Maybe it's a trade-off,
  worth considering, though - or there is a better way.

As discussed with @Hoverbear and @lukesteensen.

[RegexSet::matches]: https://docs.rs/regex/1.3.7/regex/struct.RegexSet.html#method.matches

### CODE REVIEW

**Comment by binarylogic** [comment]
Very nice! Thanks for your work on this. I'd like to see if we can deprecate the previous `regex` option without breaking backward compatibility. If you don't want to spend time doing that we can get someone on the team to implement that.

**Comment by mre** [comment]
> I'd like to see if we can deprecate the previous regex option without breaking backward compatibility.

Good idea! We could print a warning when `vector` boots, e.g.

```
⚠️ Usage of `regex` is deprecated and will be removed with version 1.0.
Please upgrade your config to use `regexes` instead like so:
`regexes = ['<pattern_here>']`
```

The `<pattern_here>` part would be the actual pattern from the user's config to make upgrading very easy.
Even better, we could provide a `vector --fix` command (similar to [rustfix](https://github.com/rust-lang-nursery/rustfix/)), that would automatically update the configuration to the latest version.

**Comment by binarylogic** [comment]
Agree! I say we start with a single-line `warn` log message and then consider the `--fix` option with something more general, like #1037. As you can see in #1037, we'd like to version our configuration and use that as a guide for updating. We're still discussing that, but I think that would be the best approach. Curious what you think!

**Comment by mre** [comment]
@binarylogic sounds good!
I've added the warning now and it looks okay to me when I run `vector --config vector.toml`:

```
May 04 15:18:39.938  INFO vector: Vector is starting. version="0.9.1" git_version="v0.9.0-105-g57dd046" released="Mon, 04 May 2020 13:14:03 +0000" arch="x86_64"
May 04 15:18:39.944  WARN vector::transforms::regex_parser: Usage of `regex` is deprecated and will be removed in a future version. Please upgrade your config to use `regexes` instead: `regexes = ['^/appdata/nomad/alloc/(?P<alloc_id>\w[^/]*?)/alloc/logs/(?P<container_name>.*?)\.(?P<logtype>\w+).*']`
```

Let me know if the wording should be changed.

The CI pipeline is currently failing, but it doesn't look like it's related to the changes in this PR.
Before merging this, I'd propose to run a benchmark to check for performance regressions.

File: src/transforms/regex_parser.rs [hunk 1]
```diff
@@ -68,16 +73,56 @@ impl RegexParser {
...
+                );
+            }
+            (None, _) => config.regexes.clone(),
+            (Some(regex), 0) => {
+                // Still using the old `regex` syntax.
+                // Printing a warning and wrapping input in a `vec`.
+                warn!(
+                    "Usage of `regex` is deprecated and will be removed in a future version. \
+                     Please upgrade your config to use `regexes` instead: \
+                     `regexes = ['{}']`",
```
Comment (mre) [inline src/transforms/regex_parser.rs:90]: Shall we add a link to the [regex transform documentation](https://vector.dev/docs/reference/transforms/regex_parser/) here?

File: src/transforms/regex_parser.rs [hunk 2]
```diff
@@ -51,14 +55,15 @@ impl TransformConfig for RegexParserConfig {
...
 pub struct RegexParser {
-    regex: Regex,
+    regexset: RegexSet,
+    regexes: Vec<Regex>,
     field: Atom,
     drop_field: bool,
     drop_failed: bool,
     target_field: Option<Atom>,
     overwrite_target: bool,
     capture_names: Vec<(usize, Atom, Conversion)>,
```
[the remark this answers is not in the document]
Comment (mre) [reply src/transforms/regex_parser.rs:65]: That's by design. Mentioned it in the design decisions section above, but we can totally change that. :)

File: src/transforms/regex_parser.rs [hunk 3]
```diff
@@ -137,11 +194,22 @@ impl Transform for RegexParser {
...
 
         if let Some(value) = &value {
-            if self
-                .regex
-                .captures_read(&mut self.capture_locs, &value)
-                .is_some()
-            {
+            let regex_id = self.regexset.matches(&value).into_iter().nth(0);
+            let (id, regex) = match regex_id {
+                Some(id) => (id, self.regexes[id].clone()), // TODO: Avoid clone
```
[the remark this answers is not in the document]
Comment (mre) [reply src/transforms/regex_parser.rs]: True, done.

File: src/transforms/regex_parser.rs [hunk 4]
```diff
@@ -137,11 +194,22 @@ impl Transform for RegexParser {
...
+                    emit!(RegexFailedMatch { value });
+                    if self.drop_failed {
+                        return None;
+                    } else {
+                        return Some(event);
+                    }
+                }
+            };
+
+            let mut capture_locs = self.capture_locs[id].clone(); // TODO: Avoid clone
```
[the remark this answers is not in the document]
Comment (mre) [reply src/transforms/regex_parser.rs]: True, done.

File: src/transforms/regex_parser.rs [hunk 5]
```diff
@@ -391,4 +463,20 @@ mod tests {
...
         assert_eq!(log[&"time".into()], Value::Float(6789.01));
     }
+
+    #[test]
+    fn supports_multiple_patterns() {
+        let log = do_transform(
+            "bar=1234",
+            r#"[
+                'foo=(?P<foo_status>\d+)',
+                'bar=(?P<bar_status>\d+)',
```
[the remark this answers is not in the document]
Comment (mre) [reply src/transforms/regex_parser.rs]: Good point. Extended the test now. Aaaand it's failing. 😅 
Can't see the error right now. Will look at it again tomorrow - unless you can find it earlier. 😉

File: src/transforms/regex_parser.rs [hunk 6]
```diff
@@ -16,7 +16,11 @@ use string_cache::DefaultAtom as Atom;
 #[derivative(Default)]
 #[serde(default, deny_unknown_fields)]
 pub struct RegexParserConfig {
-    pub regex: String,
+    /// Deprecated. Use `regexes` instead.
+    /// See #2469.
+    /// TODO: Remove at a future point in time.
+    regex: Option<String>,
+    pub regexes: Vec<String>,
```
[the remark this answers is not in the document]
Comment (mre) [reply src/transforms/regex_parser.rs:23]: I'm all for it. In my book, `patterns` is the better name. If @binarylogic and @Hoverbear agree I'll happily change it.

File: src/transforms/regex_parser.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (mre) [reply src/transforms/regex_parser.rs:65]: @lukesteensen is that fine with you?

File: src/transforms/regex_parser.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (mre) [reply src/transforms/regex_parser.rs]: Test fixed and extended. Let me know if we'd like to test other edge-cases here that I might have forgotten. 👍

**Comment by mre** [comment]
Found some time to fix the tests. The PR should be good to go now.
@lukesteensen @binarylogic can you have another look please? 😊 

There's a caveat in the implementation:
Say your input is `1234 235.42 true` and the regexes are

```toml
regexes = [
  '(?P<id>\d+)',
  '(?P<id>\d+) (?P<time>[\d.]+) (?P<check>\S+)',
]
```

then the *first* pattern will match. That's because it matches the pattern anywhere in the input. It's expected behavior for regular expressions of course, but in the case of multiple patterns, this can be confusing. (At least I was confused by that, hence the failing test.)

The proper solution is to capture the entire line like so:

```toml
regexes = [
  '^(?P<id>\d+)$',
  '^(?P<id>\d+) (?P<time>[\d.]+) (?P<check>\S+)$',
]
```

This works as "intended" and the second pattern matches.
I wonder if we should document this somehow or if it's obvious.
Alternatively, we could add another `exhaustive` option or so, which would enforce that the full pattern has to be matched. I'd argue that this would make things unnecessarily complex, though.

**Comment by Hoverbear** [comment]
@mre I think a little warning in the docs is plenty. :) I don't think an extra option is necessary.

File: src/transforms/regex_parser.rs (hunk 1 above)
Comment (Hoverbear) [reply src/transforms/regex_parser.rs:90]: Seems like a good idea!

File: src/transforms/regex_parser.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (lukesteensen) [reply src/transforms/regex_parser.rs:65]: Understood, that makes sense!

**Comment by Hoverbear** [comment]
Fine with me!

**Comment by mre** [comment]
Okay renamed `regexes` to `patterns` now and signed off the commits.
Also added a link to the docs in case a config is still using the old `regex` field.
The only missing issue from my side is the failing `make generate`. It seems to fail on a seemingly unrelated change, but I might be wrong:

```
check-blog_1                                      | `/` is not writable.
check-blog_1                                      | Bundler will use `/tmp/bundler20200512-1-chzyp41' as your home directory temporarily.
check-blog_1                                      | ---> The resulting hash from the `/.meta/**/*.toml` files failed
check-blog_1                                      |      validation against the following schema:
check-blog_1                                      |      
check-blog_1                                      |          /.meta/schema/meta.json
check-blog_1                                      |      
check-blog_1                                      |      The errors include:
check-blog_1                                      |      
check-blog_1                                      |          * The value at `/sources/vector/options/tls/children/verify_certificate/kubernetes` failed validation for `/definitions/field/additionalProperties`, reason: `schema`
```

@binarylogic is that something you could help me with?

**Comment by Hoverbear** [comment]
@mre It looks like the problem exists on master as well
image

**Comment by lukesteensen** [comment]
I think that's something we can sort on master, so I'm going to go ahead and merge this as-is.

Thanks again @mre! We're super appreciative of the time and effort you've put in here to make Vector better.
